network/src/peers/peer_book.rs

```diff
@@ -14,503 +14,263 @@
-    fn from(book: SerializedPeerBook) -> Self {
-        PeerBook {
-            disconnected_peers: RwLock::new(
-                book.0
-                    .into_iter()
-                    .filter(|info| !info.address().ip().is_loopback())
-                    .map(|info| (info.address(), info))
-                    .collect(),
-            ),
-            ..Default::default()
-        }
-    }
-}
+use tokio::{net::TcpStream, sync::mpsc};
 
 ///
 /// A data structure for storing the history of all peers with this node server.
 ///
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct PeerBook {
-    /// The map of the addresses currently being handshaken with.
-    connecting_peers: RwLock<HashSet<SocketAddr>>,
-    /// The map of connected peers to their metadata.
-    connected_peers: RwLock<HashMap<SocketAddr, PeerInfo>>,
-    /// The map of disconnected peers to their metadata.
-    disconnected_peers: RwLock<HashMap<SocketAddr, PeerInfo>>,
+    disconnected_peers: MpmcMap<SocketAddr, Peer>,
+    connected_peers: MpmcMap<SocketAddr, PeerHandle>,
+    pending_connections: Arc<AtomicU32>,
+    peer_events: mpsc::Sender<PeerEvent>,
 }
 
-impl PeerBook {
-    // TODO (howardwu): Implement manual serializers and deserializers to prevent forward breakage
-    //  when the PeerBook or PeerInfo struct fields change.
-    ///
-    /// Returns an instance of `PeerBook` from the given storage object.
-    ///
-    /// This function fetches a serialized peer book from the given storage object,
-    /// and attempts to deserialize it as an instance of `PeerBook`.
-    ///
-    /// If the peer book does not exist in storage or fails to deserialize properly,
-    /// returns a new instance of PeerBook.
-    ///
-    #[inline]
-    pub fn load<T: Transaction, P: LoadableMerkleParameters, S: Storage>(storage: &Ledger<T, P, S>) -> Self {
-        // Fetch the peer book from storage.
-        match storage.get_peer_book() {
-            // Attempt to deserialize it as a peer book.
-            Ok(Some(serialized_peer_book)) => match bincode::deserialize::<SerializedPeerBook>(&serialized_peer_book) {
-                Ok(peer_book) => PeerBook::from(peer_book),
-                _ => Default::default(),
-            },
-            _ => Default::default(),
+// to avoid circular reference to peer_events
+struct PeerBookRef {
+    disconnected_peers: MpmcMap<SocketAddr, Peer>,
+    connected_peers: MpmcMap<SocketAddr, PeerHandle>,
+    pending_connections: Arc<AtomicU32>,
+}
+
+impl PeerBookRef {
+    // gets terminated when sender is dropped from PeerBook
+    async fn handle_peer_events(self, mut receiver: mpsc::Receiver<PeerEvent>) {
+        while let Some(event) = receiver.recv().await {
+            match event.data {
+                PeerEventData::Connected(handle) => {
+                    self.pending_connections.fetch_sub(1, Ordering::SeqCst);
+                    if let Some(old_peer) = self.connected_peers.insert(event.address, handle).await {
```

Comment:
can we even run into a double-connection like this? we should at least `WARN` when this happens, as it's undesirable

Reply:
it can happen in some (mostly) synthetic cases like two peers send outbound connections to each other at the exact same time. will add a log

Reply:
also hung connections we haven't timed out yet, but the peer reconnects